### DISCUSSION TITLE
Using `rocks.nvim` with Nix and `home-manager`

### DISCUSSION THREAD

**Discussion by florentc** [lumen-oss/rocks.nvim#229] [Q&A] [answered]
I use home-manager on NixOS and currently manage my Neovim configuration using a traditional `init.lua` and `lazy.nvim` instead of plugins from `pkgs.vimPlugins` in the `plugins` section of the Neovim home-manager module. If a dependency is required by a plugin, I add it to `extraPackages` (for example `ripgrep` is useful to Telescope, `gcc` to compile treesitter parsers for `nvim-treesitter`, and so on).

I would like to give `rocks.nvim` a go. This seems like a good compromise: although still a bit redundant with Nix as another package management solution, it doesn't have the downsides of lazy.nvim.

**My open question is: how do you work with Nix, home-manager, and `rocks.nvim`, what advice would you share?**

I see there is a `luajitPackages.rocks-nvim` package on nixpkgs and this repo has a flake so I am hopeful this should all play nicely together without any complex installation procedure :+1:. Is there any guide for configuration with nix/home-manager?
I tried to add `luajitPackages.rocks-nvim` to `plugins` in the Neovim home-manager module but this does not expose the `:Rocks` command or anything in neovim (no issue with some other plugins from `pkgs.vimPlugins`).

**Comment by mrcjkb** [comment, marked as answer]
Hey :wave: 

The package in nixpkgs is currently a bit outdated, and I use the flake output mainly for testing during development.
In principle, you could use the `neovim-with-rocks` derivation (which I use for testing) as a basis for setting up rocks.nvim with a nix/home-manager managed neovim.

But here are some caveats to be aware of when using the nix derivation:

- rocks.nvim is designed to manage itself, so it adds a "rocks.nvim" plugin entry to the rocks.toml.
  If you don't remove that entry, it'll install the rocks.nvim lua package, and you'll have two conflicting installations.
  One in the nix store, and one managed by rocks.nvim.
- If you use nix to manage the rocks.nvim package, and remove the entry from rocks.toml, it will still install rocks.nvim from luarocks if you install a module that depends on it (e.g. rocks-git.nvim or rocks-config.nvim). So in this case, you would need to manage all rocks.nvim extensions with nix.

An alternate solution would be to add `lua51Packages.luarocks`, and configure Neovim to bootstrap rocks.nvim if it's not installed.

Something along the lines of this[^1]:

[^1]: Note that the code snippet is meant to be used in a nix config, as it injects `"${pkgs.lua51Packages.luarocks}/bin/luarocks"`

```lua
local rocks_config = {
    rocks_path = vim.fn.stdpath("data") .. "/rocks",
    luarocks_binary = "${pkgs.lua51Packages.luarocks}/bin/luarocks", 
}

vim.g.rocks_nvim = rocks_config

local luarocks_path = {
    vim.fs.joinpath(rocks_config.rocks_path, "share", "lua", "5.1", "?.lua"),
    vim.fs.joinpath(rocks_config.rocks_path, "share", "lua", "5.1", "?", "init.lua"),
}
package.path = package.path .. ";" .. table.concat(luarocks_path, ";")

local luarocks_cpath = {
    vim.fs.joinpath(rocks_config.rocks_path, "lib", "lua", "5.1", "?.so"),
    vim.fs.joinpath(rocks_config.rocks_path, "lib64", "lua", "5.1", "?.so"),
}
package.cpath = package.cpath .. ";" .. table.concat(luarocks_cpath, ";")

if not vim.uv.fs_stat(rocks_config.rocks_path) then
  vim.system({
    rocks_config.luarocks_binary,
    "install",
    "--tree",
    rocks_config.rocks_path, 
    "--server='https://nvim-neorocks.github.io/rocks-binaries/'",
    "--lua-version=5.1",
    "rocks.nvim",
    }):wait()
end

vim.opt.runtimepath:append(vim.fs.joinpath(rocks_config.rocks_path, "lib", "luarocks", "rocks-5.1", "rocks.nvim", "*"))
```

Some more important info:

- You will need an up-to-date neovim-nightly package (e.g. from the [neovim-nightly-overlay](https://github.com/nix-community/neovim-nightly-overlay).

Since I use NixOS myself, I am definitely very interested in creating a module or setup that makes it easier to use rocks.nvim with NixOS.
But right now, my priorities are polishing it for "average" users.
It would definitely be great to have early adopters on NixOs/home-manager though, so I can get feedback :smile: 

You probably won't need gcc for tree-sitter parsers, because rocks.nvim fetches pre-compiled parsers from our rocks-binaries server.

**Reply by c3n21** [reply]
Hi! Joining the discussion here as I'm a fellow (and newbie) Nix user myself and this is exactly what I'm trying to do in my `init.lua`, the only difference being I'm using the bootstrap.lua script provided in the README which looks similar.

```lua
 do
     -- Specifies where to install/use rocks.nvim
     local install_location = vim.fs.joinpath(vim.fn.stdpath('data'), 'rocks')
     local luarocks_path =  vim.fn.exepath('luarocks')
     print(luarocks_path)
 
     -- Set up configuration options related to rocks.nvim (recommended to leave as default)
     local rocks_config = {
         rocks_path = vim.fs.normalize(install_location),
         luarocks_binary = luarocks_path
     }
 
     vim.g.rocks_nvim = rocks_config
 
     -- Configure the package path (so that plugin code can be found)
     local luarocks_path = {
         vim.fs.joinpath(rocks_config.rocks_path, 'share', 'lua', '5.1', '?.lua'),
         vim.fs.joinpath(rocks_config.rocks_path, 'share', 'lua', '5.1', '?', 'init.lua'),
     }
     package.path = package.path .. ';' .. table.concat(luarocks_path, ';')
 
     -- Configure the C path (so that e.g. tree-sitter parsers can be found)
     local luarocks_cpath = {
         vim.fs.joinpath(rocks_config.rocks_path, 'lib', 'lua', '5.1', '?.so'),
         vim.fs.joinpath(rocks_config.rocks_path, 'lib64', 'lua', '5.1', '?.so'),
     }
     package.cpath = package.cpath .. ';' .. table.concat(luarocks_cpath, ';')
 
     -- Load all installed plugins, including rocks.nvim itself
     vim.opt.runtimepath:append(
         vim.fs.joinpath(rocks_config.rocks_path, 'lib', 'luarocks', 'rocks-5.1', 'rocks.nvim', '*')
     )
 end
 
 print(vim.inspect(vim.g.rocks_nvim))
 
 -- If rocks.nvim is not installed then install it!
 if not pcall(require, 'rocks') then
     local rocks_location = vim.fs.joinpath(vim.fn.stdpath('cache'), 'rocks.nvim')
 
     if not vim.uv.fs_stat(rocks_location) then
         -- Pull down rocks.nvim
         vim.fn.system({
             'git',
             'clone',
             '--filter=blob:none',
             'https://github.com/nvim-neorocks/rocks.nvim',
             rocks_location,
         })
     end
 
     -- If the clone was successful then source the bootstrapping script
     assert(vim.v.shell_error == 0, 'rocks.nvim installation failed. Try exiting and re-entering Neovim!')
 
     vim.cmd.source(vim.fs.joinpath(rocks_location, 'bootstrap.lua'))
 
     vim.fn.delete(rocks_location, 'rf')
 end
```

and `vim.fn.exepath('luarocks')`returns me `/home/zhifan/.nix-profile/bin/luarocks` (I installed `lua51Packages.luarocks` through home-manager and only for my current user)

but I still get `sh configure` complaining about not finding Lua and Lua's headers.

How can I fix this?

This is the error message I get (I did `print(vim.inspect(sc))` inside of `bootstrap.lua`) 

`
{
  code = 1,
  signal = 0,
  stderr = "",
  stdout = "\n\27[1;36mConfiguring LuaRocks version dev...\27[0m\n\n\27[1;31mLua 5.1 interpreter not found\27[0m in $PATH\nYou may want to use the flags \27[1m
--with-lua\27[0m, \27[1m--with-lua-bin\27[0m and/or \27[1m--lua-version\27[0m\nRun \27[1m./configure --help\27[0m for details.\n\n\27[1;31mconfigure failed.\27
[0m\n\n"
}
`

I noticed that if I install through the installer  and flag "setup luarocks" as false it will just clone `rocks.nvim` and it seems to work properly.

Maybe in the bootstrap script we should just ignore the piece where it installs luarocks locally and use the one on $PATH?

EDIT:

forgot to mention that I am using https://github.com/nix-community/neovim-nightly-overlay

**Reply by mrcjkb** [reply]
@c3n21 the bootstrap and installer scripts aren't meant for use on NixOS, as the luarocks configure script (not managed by us) expects a Linux filesystem hierarchy standard (which NixOS doesn't comply with).

We have an open issue for detecting properly configured luarocks installations: https://github.com/nvim-neorocks/rocks.nvim/issues/218

**Reply by c3n21** [reply]
thanks for the reply!

So for now I could just "imperatively" install rocks.nvim by deselecting "setup luarocks" right in the installer right?

**Reply by mrcjkb** [reply]
Yes, that should work.

**Reply by teto** [reply]
I am interested in handling this via nix. Would it make sense to add an option to disallow rocks.nvim autohandling itself ? As it's a critical piece of software, even outside nix, someone may want to control himself how to update rocks.nvim

**Comment by florentc** [comment]
Thanks for your detailed answer! I wouldn't mind going the way you suggest: providing bootstrapping dependencies with nix, and having neovim fetch rocks.nvim and set itself up. This is what I am already doing with lazy.nvim (I don't fetch lazy from nixpkgs).

The README describes a very imperative approach to plugin configuration using `:Rocks install` and so on. If I am not mistaken, those commands mutate the `rocks.toml` file. Can I manage this `rocks.toml` with home-manager instead? The deployed `rocks.toml` would be read-only and maybe it is not supposed to. Can the plugin still work in those conditions (apart from `:Rocks install` and the like which would fail of course).

**Reply by mrcjkb** [reply]
Yes, you could use `builtins.toTOML` to create the rocks.toml from a nix attrset, and point `vim.g.rocks_nvim.config_path` to your generated rocks.toml (or use home-manager's `xdg.configFile`).

The commands that try to write to `rocks.toml` are `:Rocks install`, `:Rocks update` and `:Rocks prune`.
You can still use `:Rocks sync`, which only reads from `rocks.toml`.

**Comment by AlphabetsAlphabets** [comment]
Hi, @mrcjkb. I'm a bit lost about your instruction. I have included the bootstrapping code you mentioned above and I also have these packages installed.

```
      lua51Packages.lua lua51Packages.luarocks unstable.luajitPackages.rocks-nvim
```

But when I install rocks through the script and paste the code you provided and start neovim I run into this error.

```lua
Error detected while processing /home/star/.local/share/nvim/rocks/lib/luarocks/rocks-5.1/rocks.nvim/2.25.
1-1/plugin/rocks.lua:
E5113: Error while calling lua chunk: ...unwrapped-7f08477/share/nvim/runtime/lua/vim/_system.lua:244: ENO
ENT: no such file or directory
stack traceback:
        [C]: in function 'error'
        ...unwrapped-7f08477/share/nvim/runtime/lua/vim/_system.lua:244: in function 'spawn'
        ...unwrapped-7f08477/share/nvim/runtime/lua/vim/_system.lua:335: in function 'system'
        ...r/.local/share/nvim/rocks/share/lua/5.1/rocks/loader.lua:7: in function 'get_luarocks_lua_dir_f
rom_luarocks'
        ...r/.local/share/nvim/rocks/share/lua/5.1/rocks/loader.lua:18: in function 'enable'
        .../luarocks/rocks-5.1/rocks.nvim/2.25.1-1/plugin/rocks.lua:29: in main chunk
```

**Reply by mrcjkb** [reply]
Hey 👋 

It looks like rocks.nvim can't find the luarocks binary. 
Check your `vim.g.rocks_nvim.luarocks_binary`.

**Reply by AlphabetsAlphabets** [reply]
This is the output `${pkgs.lua51Packages.luarocks}/bin/luarocks`

**Reply by mrcjkb** [reply]
`${pkgs.lua51Packages.luarocks}` in my snippet assumes you've defined the init.lua in nix. 
The resulting `vim.g.rocks_nvim.luarocks_binary` should point to a luarocks binary in the nix store.

**Reply by AlphabetsAlphabets** [reply]
When you say 

> ${pkgs.lua51Packages.luarocks} in my snippet assumes you've defined the init.lua in nix.

Do you mean the init.lua is defined inside `configuration.nix`?

> The resulting vim.g.rocks_nvim.luarocks_binary should point to a luarocks binary in the nix store.

In my case, I need to find out the luarocks binary and point to it manually? If so, how can I do that?

**Reply by mrcjkb** [reply]
> Do you mean the init.lua is defined inside configuration.nix?

Yes, see [this article](https://wiki.nixos.org/wiki/Neovim).

> In my case, I need to find out the luarocks binary and point to it manually? If so, how can I do that?

If you have luarocks in your `systemPackages`, you should be able to remove the `luarocks_binary` field from your config. It defaults to `"luarocks"`, expecting it to be on the `PATH`.

**Reply by AlphabetsAlphabets** [reply]
Removing the `luarocks_binary` field from the boostrapping code fixed the original error but now I have a new one.

```lua
Error detected while processing /home/star/.local/share/nvim/rocks/lib/luarocks/rocks-5.1/rocks.nvim/2.7.2-1/plugin/rocks.lua:
E5113: Error while calling lua chunk: error loading module 'toml' from file '/home/star/.local/share/nvim/rocks/lib/lua/5.1/toml.so':
        libstdc++.so.6: cannot open shared object file: No such file or directory
stack traceback:
        [C]: at 0x7f4ec5765080
        [C]: in function 'require'
        ...share/nvim/rocks/share/lua/5.1/rocks/config/internal.lua:49: in function 'get_rocks_toml'
        ...share/nvim/rocks/share/lua/5.1/rocks/config/internal.lua:69: in function 'get_user_rocks'
        .../.local/share/nvim/rocks/share/lua/5.1/rocks/runtime.lua:161: in function 'source_start_plugins'
        ...b/luarocks/rocks-5.1/rocks.nvim/2.7.2-1/plugin/rocks.lua:22: in main chunk
```

Looks like a file is missing. While the `toml.so` does exist. In the same folder where `toml.so` is supposed to be, the `rocks` folder doesn't exist.

```
star@platinum ~/.l/s/n/r/l/l/5.1> pwd
/home/star/.local/share/nvim/rocks/lib/lua/5.1
star@platinum ~/.l/s/n/r/l/l/5.1> ls
fzy_native.so  toml_edit.so  toml.so
```

**Reply by mrcjkb** [reply]
It looks like this is the actual error:

```
libstdc++.so.6: cannot open shared object file
```

But rocks.nvim no longer depends on toml.lua (which comes with toml.so). So updating to the latest version should fix that.

**Reply by AlphabetsAlphabets** [reply]
I see, I didn't notice that the version was `2.1.2` lol. Thanks :D

**Comment by florentc** [comment]
Out of curiosity @mrcjkb, as a Nix and Neovim user yourself, what is your personal setup for your own day to day neovim instance? Do you use `rocks.nvim` in this way? And when developping `rocks.nvim`, how do you test and experiment in an "average" user environnement? You have another "rocks-ified" nvim besides your personal nvim?

You said earlier that the average user is your current target, but I am also curious to know what you envision would be the longer term approach to ease usage of `rocks.nvim` with Nix, if this were on your roadmap someday.

**Reply by mrcjkb** [reply]
Hey 👋 

Sorry for the late reply, I just saw this now. 
I don't daily drive rocks.nvim myself (yet). My [personal config](https://github.com/mrcjkb/nvim) uses nix to manage plugins, and is what spawned [kickstart-nix.nvim](https://github.com/nix-community/kickstart-nix.nvim).

My personal motivation for contributing to rocks.nvim is the fact that nixpkgs uses luarocks to build neovim plugins that are available on luarocks.org (see @teto's [blog post](https://teto.github.io/posts/2022-06-22-neovim-plugin-luarocks-2.html)).

For manual testing, I use the `#neovim-with-rocks` flake output of this repo (which has `$NVIM_APPNAME` set to `"nvimrocks"`).

For ease of use with Nix, I have some things in mind (but not fully thought out yet):

- A NixOS/home-manager module that lets you enable rocks.nvim extensions (like rocks-git.nvim, rocks-config.nvim, etc.).
  *If installing rocks.nvim with nix, any package that has rocks.nvim as a dependency also has to be managed with nix, otherwise luarocks will install another (potentially conflicting) rocks.nvim locally.*
-  A `rocks-nix.nvim` extension (equivalent to `rocks-git.nvim`) that can be used to manage plugins using nix, and tell rocks.nvim where to find them. With such an extension, you could still use rocks.nvim to install plugins from luarocks (e.g. for trying them out) and then migrate them to your nix config.

**Reply by teto** [reply]
I would like to write more extensively about nix + rocks but until I find the time, I configure it
```
local luarocks_config_fn = assert(loadfile(vim.fn.stdpath('config') .. '/luarocks-config-generated.lua'))
vim.g.rocks_nvim = {
    luarocks_binary = nix_deps.luarocks_executable,
    luarocks_config = luarocks_config_fn(), -- , luarocks_config = nil
    _log_level = vim.log.levels.TRACE,
    lazy = true, -- for cleaner logs
}
```
where nix_deps is an array with nix generated strings https://github.com/teto/home/blob/a4dfe4a01203be09855afe1422f0359ab0824eaa/hm/profiles/neovim.nix#L110

**Reply by teto** [reply]
I've dumped my config at https://github.com/nvim-neorocks/rocks.nvim/wiki/Install-with-nix . Hopefully we can make rocks+nix more plugandplay in the short future.

**Comment by insipx** [comment]
For those looking for a concrete example, this is how I configured my [neovim](https://github.com/insipx/home.nix/blob/main/home-manager/configure-neovim.nix)

I import that into my home-manager config.

Then run the imperative install command, 

`nvim -u NORC -c "source https://raw.githubusercontent.com/nvim-neorocks/rocks.nvim/master/installer.lua"` 

making sure to set `install luarocks` to 'false'

Not a perfect solution but just pre-pends the configuration above to init.lua.
